core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -403,9 +403,26 @@ static TableMetadata fromJson(FileIO io, InputFile file, JsonNode node) {
       }
     }
 
-    return new TableMetadata(file, formatVersion, uuid, location,
-        lastSequenceNumber, lastUpdatedMillis, lastAssignedColumnId, currentSchemaId, schemas, defaultSpecId, specs,
-        lastAssignedPartitionId, defaultSortOrderId, sortOrders, properties, currentVersionId,
-        snapshots, entries.build(), metadataEntries.build());
+    return TableMetadata.builder()
```

Comment:
What is the benefit of using the builder here instead of instantiating `TableMetadata` directly? Looks like it made the method too long! I don't think that this needs to use the builder because there isn't anything to validate or change.

Reply:
I think the benefit of a builder for this class would be using it to accumulate state as it is read, rather than keeping the local variables.

For example, rather than calling `snapshots.add` and then calling `withSnapshots(snapshots)`, the builder could help by exposing a `addSnapshot` method. But before adding that, I think it's worth considering what the purpose of the builder is, and I don't think that it is to make parsing easier.

To me, the purpose of the builder is to make updating table metadata easier. That's what I initially introduced the high-level methods for, like `TableMetadata.replaceCurrentSnapshot`. I think a builder pattern works better for implementing those methods since you can make multiple changes in the same builder instead of building a new `TableMetadata` and calling a different high-level method on it.

So in the end, I don't think that I would update this just yet like I said in my first comment. And I would probably change the API of the builder in general. If you want to use a builder here, then maybe two builders are appropriate. One for building from existing table metadata and one for just constructing the `TableMetadata` objects.